### ISSUE TITLE
Extracting the raw DER signature from a PE file is unexpectedly slow

### ISSUE THREAD

**Issue by jhhcs** [lief-project/LIEF#832]
**Description:** I am using the Python bindings of LIEF. When extracting the raw DER signature of a PE file, like so:
```python
pe = lief.PE.parse(BytesIO(data))
for signature in pe.signatures:
    yield signature.raw_der
```
it takes roughly 2.3 seconds on my machine. I also have code based on `pefile` which does this in 0.03 seconds. I am trying to replace pefile entirely with LIEF in this project, but the performance difference is an issue here. I assume that the reason for this is that LIEF performs a lot of parsing and checking that is not necessary for extracting the raw DER data.

**Expected behavior:** I would expect the extraction of the raw DER data to take only milliseconds.

**Environment (please complete the following information):**
 - System and Version: Windows 10
 - Target format: PE
 - LIEF commit version: 0.12.3-39115d10

**Comment by romainthomas** [comment]
That's an issue indeed. Could you attach the signature(s) that make the parsing slow?

**Comment by jhhcs** [comment]
Firstly, my sample code in the initial post is not quite accurate. I think it's best to explain what I did so you get the entire user pespective:
- First, I tried passing a `bytes` object to `lief.PE.parse`. This unexpectedly spammed my terminal with the entire content of the binary, I assume as part of an error message because it reated the `bytes` object like a filename string (it should really only do that for `str` objects).
- I then tried to use `io.BytesIO`, but this failed with the following error:
   ```
   Traceback (most recent call last):
     File "W:\temporary\sig.py", line 40, in <module>
       pe = lief.PE.parse(io.BytesIO(data))
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   AttributeError: '_io.BytesIO' object has no attribute 'raw'
   ```
- I then studied [the documentation](https://lief-project.github.io/doc/latest/api/python/pe.html) and discovered that LIEF supports only a list of integers as input (somewhat related to #833) and I started using `list(data)` instead. This is the first thing that worked, and _this was very slow_.
-  I have since found out that using a custom wrapper around `io.BytesIO` also works, and it is a bit faster.
 
Next, I will attach the sample I used for testing, but please be advised that it is malware, so I will put this here:

**⚠ WARNING: The [attached ZIP file](https://github.com/lief-project/LIEF/files/10175827/testfile.bin.zip) contains a Windows malware sample. Do not execute it under any circumstances. The archive is protected with the password "infected".**

I wrote the following test snippet which uses a different wrapper. It requires a wrapper in any case because LIEF seems to access the property `raw` of the input io object, which does not exist on `io.BytesIO` objects by default:
```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import time
import io
import hashlib
import lief
import pefile

data = open('testfile.bin', 'rb').read()

t_start = time.process_time()
sd = pefile.DIRECTORY_ENTRY['IMAGE_DIRECTORY_ENTRY_SECURITY']
pe = pefile.PE(data=data, fast_load=True)
pe.parse_data_directories(directories=[sd])
sd_info = pe.OPTIONAL_HEADER.DATA_DIRECTORY[sd]
sd_pos = sd_info.VirtualAddress + 8
sd_end = sd_pos + sd_info.Size
sd_hash = hashlib.sha256(data[sd_pos:sd_end]).hexdigest()
t_stop = time.process_time()
print(F'{sd_hash} {t_stop-t_start:5.3f} pefile fast load')

t_start = time.process_time()
sd = pefile.DIRECTORY_ENTRY['IMAGE_DIRECTORY_ENTRY_SECURITY']
pe = pefile.PE(data=data, fast_load=False)
sd_info = pe.OPTIONAL_HEADER.DATA_DIRECTORY[sd]
sd_pos = sd_info.VirtualAddress + 8
sd_end = sd_pos + sd_info.Size
sd_hash = hashlib.sha256(data[sd_pos:sd_end]).hexdigest()
t_stop = time.process_time()
print(F'{sd_hash} {t_stop-t_start:5.3f} pefile default')

class BytesIO(io.BytesIO):
    @property
    def raw(self):
        return self
    def readall(self):
        return self.read()

t_start = time.process_time()
pe = lief.PE.parse(BytesIO(data))
for signature in pe.signatures:
    sd_data = bytearray(signature.raw_der)
    sd_hash = hashlib.sha256(sd_data).hexdigest()
    t_stop = time.process_time()
    print(F'{sd_hash} {t_stop-t_start:5.3f} lief')
    break
```
It produces the following output in a (representative) test run:
```
886d1f4dcc495ce50c2f58d7b2eff3a28daecc3f3c4c84193bad33a5b8f81935 0.000 pefile fast load
886d1f4dcc495ce50c2f58d7b2eff3a28daecc3f3c4c84193bad33a5b8f81935 0.359 pefile default
886d1f4dcc495ce50c2f58d7b2eff3a28daecc3f3c4c84193bad33a5b8f81935 0.781 lief io object
886d1f4dcc495ce50c2f58d7b2eff3a28daecc3f3c4c84193bad33a5b8f81935 2.078 lief list
```

**Comment by romainthomas** [comment]
Thank you for the clarification @jhhcs. If I understand correctly fixing the `ByteIO` interface in LIEF should ok right?

**Comment by huettenhain** [comment]
(Same person here, just a different account) I would probably have to say "no", at least if it is up to me, because even with the patched `BytesIO`, it still takes about 800 times as long as pefile with the `fast_load` option, which is what I currently use. It is also still twice as slow as having pefile parse the entire header - and I moved to `fast_load` because 0.3 to 0.4 seconds for each file was too much 😕. I want to emphasize that I can absolutely live with the answer that this can't or won't be fixed, but I had hoped that LIEF would reach a similar performance as pefile _with_ `fast_load`, because the parsing is happening in superfast C++ land 🙂.

**Comment by romainthomas** [comment]
Got it. At this point this is a design choice in LIEF to process all the PE directories compared to the `fast_load` option in pefile.

Nonetheless, the behavior could be tweaked as it is done for the Mach-O format[^1]. I'm completely ok with this feature and I think it worth it but right know I don't have *bandwidth* to implement it. It should come with some refactoring I had in mind since a while for the PE parser/builder, but I can't tell you when it will done. 

[^1]: [ParserConfig.hpp](https://github.com/lief-project/LIEF/blob/57fcff35c51d0e7f0779cc20b6c1c6196bed5631/include/LIEF/MachO/ParserConfig.hpp#L23-L49)

**Comment by jhhcs** [comment]
I really am very happy if it's on your radar at all, and I wouldn't dream of prompting you for a timeline, this is a free and open source project after all. I should also mention that LIEF is awesome, the Python bindings are exceptional for a C++ lib, and it has an amazing scope. I am happy I get to use it exactly the way it is right now, even if I have to keep pefile around for a few corner cases like this.

Depending on how you prefer to manage your issues, please feel free to close this out if a rework of the parser is on the roadmap anyway.

**Comment by romainthomas** [comment]
Thank you @huettenhain :)

I pushed [2a10494](https://github.com/lief-project/LIEF/commit/2a10494b8a5427a9e6316d1cd97af49cb00a3e6a) which fixes the io interface and I opened #839 for tracking the `fast_load` feature.
